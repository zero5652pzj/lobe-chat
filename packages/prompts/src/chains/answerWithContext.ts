import { ChatStreamPayload } from '@lobechat/types';

interface AnswerWithContext {
  context: string[];
  knowledge: string[];
  question: string;
}
export const chainAnswerWithContext = ({
  context,
  knowledge,
  question,
}: AnswerWithContext): Partial<ChatStreamPayload> => {
  const filteredContext = context.filter((c) => c.trim());
  const hasContext = filteredContext.length > 0;

  return {
    messages: [
      {
        content: hasContext
          ? `You are a helpful assistant specialized in ${knowledge.join('/')}. Your task is to answer questions based on the provided context passages.

IMPORTANT RULES:
- First, check if the context is relevant to the question topic
- If the context is about a COMPLETELY DIFFERENT topic than the question:
  * State what topic the context is about
  * Clearly state "The provided context does not contain information about [question topic]"
  * Do NOT answer using your general knowledge
- If the context is related to the question topic (even if information is limited):
  * ALWAYS use the context information as a foundation
  * You SHOULD supplement with your general knowledge to provide a complete, helpful answer
  * For "how to" questions, MUST provide practical, actionable steps combining context + your expertise
  * The context provides the topic relevance - you provide the comprehensive answer
  * Example: If context mentions "Docker is for containerization", and question is "How to deploy with Docker?", you should explain deployment steps using your knowledge
- Answer in the same language as the question
- Use markdown formatting for better readability

The provided context passages:

<context>
${filteredContext.join('\n')}
</context>

Question to answer:

${question}`
          : `You are a helpful assistant specialized in ${knowledge.join('/')}. Please answer the following question using your knowledge.

Answer in the same language as the question and use markdown formatting for better readability.

Question to answer:

${question}`,
        role: 'user',
      },
    ],
  };
};
