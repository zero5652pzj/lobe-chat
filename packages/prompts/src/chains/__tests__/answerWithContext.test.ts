import { describe, expect, it } from 'vitest';

import { chainAnswerWithContext } from '../answerWithContext';

describe('chainAnswerWithContext', () => {
  it('should generate correct payload with context and knowledge', () => {
    const result = chainAnswerWithContext({
      context: ['Context passage 1', 'Context passage 2'],
      knowledge: ['AI', 'Machine Learning'],
      question: 'What is artificial intelligence?',
    });

    expect(result).toMatchSnapshot();
  });

  it('should handle single knowledge area', () => {
    const result = chainAnswerWithContext({
      context: ['Single context'],
      knowledge: ['Technology'],
      question: 'How does it work?',
    });

    expect(result).toMatchSnapshot();
  });

  it('should handle multiple knowledge areas', () => {
    const result = chainAnswerWithContext({
      context: ['Context'],
      knowledge: ['AI', 'ML', 'NLP', 'Computer Vision'],
      question: 'Tell me about these fields',
    });

    expect(result).toMatchSnapshot();
  });

  it('should handle empty context array', () => {
    const result = chainAnswerWithContext({
      context: [],
      knowledge: ['AI'],
      question: 'What is AI?',
    });

    expect(result).toMatchSnapshot();
  });

  it('should filter out empty context strings', () => {
    const result = chainAnswerWithContext({
      context: ['Valid context', '', '  ', 'Another valid context'],
      knowledge: ['Test'],
      question: 'Test question',
    });

    expect(result).toMatchSnapshot();
  });
});
