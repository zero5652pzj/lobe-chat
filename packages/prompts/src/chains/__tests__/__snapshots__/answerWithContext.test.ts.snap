// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`chainAnswerWithContext > should filter out empty context strings 1`] = `
{
  "messages": [
    {
      "content": "You are a helpful assistant specialized in Test. Your task is to answer questions based on the provided context passages.

IMPORTANT RULES:
- First, check if the context is relevant to the question topic
- If the context is about a COMPLETELY DIFFERENT topic than the question:
  * State what topic the context is about
  * Clearly state "The provided context does not contain information about [question topic]"
  * Do NOT answer using your general knowledge
- If the context is related to the question topic (even if information is limited):
  * ALWAYS use the context information as a foundation
  * You SHOULD supplement with your general knowledge to provide a complete, helpful answer
  * For "how to" questions, MUST provide practical, actionable steps combining context + your expertise
  * The context provides the topic relevance - you provide the comprehensive answer
  * Example: If context mentions "Docker is for containerization", and question is "How to deploy with Docker?", you should explain deployment steps using your knowledge
- Answer in the same language as the question
- Use markdown formatting for better readability

The provided context passages:

<context>
Valid context
Another valid context
</context>

Question to answer:

Test question",
      "role": "user",
    },
  ],
}
`;

exports[`chainAnswerWithContext > should generate correct payload with context and knowledge 1`] = `
{
  "messages": [
    {
      "content": "You are a helpful assistant specialized in AI/Machine Learning. Your task is to answer questions based on the provided context passages.

IMPORTANT RULES:
- First, check if the context is relevant to the question topic
- If the context is about a COMPLETELY DIFFERENT topic than the question:
  * State what topic the context is about
  * Clearly state "The provided context does not contain information about [question topic]"
  * Do NOT answer using your general knowledge
- If the context is related to the question topic (even if information is limited):
  * ALWAYS use the context information as a foundation
  * You SHOULD supplement with your general knowledge to provide a complete, helpful answer
  * For "how to" questions, MUST provide practical, actionable steps combining context + your expertise
  * The context provides the topic relevance - you provide the comprehensive answer
  * Example: If context mentions "Docker is for containerization", and question is "How to deploy with Docker?", you should explain deployment steps using your knowledge
- Answer in the same language as the question
- Use markdown formatting for better readability

The provided context passages:

<context>
Context passage 1
Context passage 2
</context>

Question to answer:

What is artificial intelligence?",
      "role": "user",
    },
  ],
}
`;

exports[`chainAnswerWithContext > should handle empty context array 1`] = `
{
  "messages": [
    {
      "content": "You are a helpful assistant specialized in AI. Please answer the following question using your knowledge.

Answer in the same language as the question and use markdown formatting for better readability.

Question to answer:

What is AI?",
      "role": "user",
    },
  ],
}
`;

exports[`chainAnswerWithContext > should handle multiple knowledge areas 1`] = `
{
  "messages": [
    {
      "content": "You are a helpful assistant specialized in AI/ML/NLP/Computer Vision. Your task is to answer questions based on the provided context passages.

IMPORTANT RULES:
- First, check if the context is relevant to the question topic
- If the context is about a COMPLETELY DIFFERENT topic than the question:
  * State what topic the context is about
  * Clearly state "The provided context does not contain information about [question topic]"
  * Do NOT answer using your general knowledge
- If the context is related to the question topic (even if information is limited):
  * ALWAYS use the context information as a foundation
  * You SHOULD supplement with your general knowledge to provide a complete, helpful answer
  * For "how to" questions, MUST provide practical, actionable steps combining context + your expertise
  * The context provides the topic relevance - you provide the comprehensive answer
  * Example: If context mentions "Docker is for containerization", and question is "How to deploy with Docker?", you should explain deployment steps using your knowledge
- Answer in the same language as the question
- Use markdown formatting for better readability

The provided context passages:

<context>
Context
</context>

Question to answer:

Tell me about these fields",
      "role": "user",
    },
  ],
}
`;

exports[`chainAnswerWithContext > should handle single knowledge area 1`] = `
{
  "messages": [
    {
      "content": "You are a helpful assistant specialized in Technology. Your task is to answer questions based on the provided context passages.

IMPORTANT RULES:
- First, check if the context is relevant to the question topic
- If the context is about a COMPLETELY DIFFERENT topic than the question:
  * State what topic the context is about
  * Clearly state "The provided context does not contain information about [question topic]"
  * Do NOT answer using your general knowledge
- If the context is related to the question topic (even if information is limited):
  * ALWAYS use the context information as a foundation
  * You SHOULD supplement with your general knowledge to provide a complete, helpful answer
  * For "how to" questions, MUST provide practical, actionable steps combining context + your expertise
  * The context provides the topic relevance - you provide the comprehensive answer
  * Example: If context mentions "Docker is for containerization", and question is "How to deploy with Docker?", you should explain deployment steps using your knowledge
- Answer in the same language as the question
- Use markdown formatting for better readability

The provided context passages:

<context>
Single context
</context>

Question to answer:

How does it work?",
      "role": "user",
    },
  ],
}
`;
